presto-orc/src/main/java/com/facebook/presto/orc/CachingOrcDataSource.java

```diff
@@ -0,0 +1,165 @@
+import java.util.ListIterator;
+import java.util.Map;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+public class CachingOrcDataSource
+        implements OrcDataSource
+{
+    private static final long MAX_CACHE_LENGTH = new DataSize(8, DataSize.Unit.MEGABYTE).toBytes();
+
+    private final OrcDataSource dataSource;
+    ListIterator<StripeInformation> stripeIterator;
+
+    private long cachePosition;
+    private int cacheLength;
+    private byte[] cache;
+
+    public CachingOrcDataSource(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        this.dataSource = requireNonNull(dataSource, "dataSource is null");
+        this.stripeIterator = requireNonNull(stripes, "stripes is null").listIterator();
+    }
+
+    public static OrcDataSource wrapIfTinyStripes(OrcDataSource dataSource, List<StripeInformation> stripes)
+    {
+        if (stripes.size() < 3) {
+            return dataSource;
+        }
+        for (StripeInformation stripe : stripes) {
+            if (stripe.getDataLength() + stripe.getFooterLength() + stripe.getIndexLength() > MAX_CACHE_LENGTH) {
+                return dataSource;
+            }
+        }
+        return new CachingOrcDataSource(dataSource, stripes);
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return dataSource.getReadTimeNanos();
+    }
+
+    @Override
+    public long getSize()
+    {
+        return dataSource.getSize();
+    }
+
+    @VisibleForTesting
+    void readCacheAt(long offset)
+            throws IOException
+    {
+        // Assumption: stripes are always read in order
+        // Assumption: bytes that are not part of any stripe is never read
+        long startOffset;
+        long endOffset;
+        while (true) {
+            verify(stripeIterator.hasNext());
+            StripeInformation firstStripe = stripeIterator.next();
+            long firstStripeLength = firstStripe.getIndexLength() + firstStripe.getDataLength() + firstStripe.getFooterLength();
+            verify(firstStripeLength <= MAX_CACHE_LENGTH);
+            startOffset = firstStripe.getOffset();
+            endOffset = firstStripe.getOffset() + firstStripeLength;
+            if (endOffset > offset) {
+                verify(startOffset <= offset);
+                break;
+            }
+        }
+        while (stripeIterator.hasNext()) {
+            StripeInformation nextStripe = stripeIterator.next();
+            long nextStripeLength = nextStripe.getIndexLength() + nextStripe.getDataLength() + nextStripe.getFooterLength();
+            long newEndOffset = nextStripe.getOffset() + nextStripeLength;
+            if (newEndOffset - startOffset > MAX_CACHE_LENGTH) {
+                stripeIterator.previous();
+                break;
+            }
+            endOffset = newEndOffset;
+        }
+        cacheLength = Ints.checkedCast(endOffset - startOffset);
+        expandCacheIfNeeded(cacheLength);
```

Comment:
This is only used in one place, and it is directly tied to `cacheLength`, so it would be clear to inline it.